### PR TITLE
Make series creation scale better with number of cores

### PR DESCRIPTION
**Pull request by tomwilkie**
Adding a new user with ~15m series, saw the ingesters latency go through the roof.

Benchmarked series creation and saw it doesn't parallelise.   Made series creation scale better with number of cores (up to ~16 cores).

This is achieved by sharding many of the datastructures on the series creation path, such as the seriesMap, the inmemory index et

Graph of test run time vs parallelisation for 1 to 32 goroutines.  Test was run on a 32 vCPU-optimised droplet.

screen shot 2018-08-08 at 15 56 34

### CODE REVIEW

File: pkg/ingester/index.go [hunk 1]
```diff
@@ -40,56 +51,73 @@ func (i *invertedIndex) add(metric model.Metric, fp model.Fingerprint) {
...
-			toIntersect = merge(toIntersect, fps)
-		} else {
-			for value, fps := range values {
-				if matcher.Matches(string(value)) {
-					toIntersect = merge(toIntersect, fps)
+	result := []model.Fingerprint{}
+
+outer:
+	for i := range ii.shards {
+		ii.shards[i].mtx.RLock()
```
Comment (leth) [inline pkg/ingester/index.go]: Is it worth assigning `ii.shards[i]` to something?

**Comment by tomwilkie** [comment]
> How did you pick the shard counts? Have you done any experimenting with tuning them?

A little thought, but not much:

- `metricCounterShards = 128` - only ever read or written by index, so little downside in having many
- `const indexShards = 32` - writes hit read on shards, reads all - so can't be too big or queries get bigger
- `const seriesMapShards = 128` - only ever  read or written by index, so little downside in having many

File: pkg/ingester/index.go (hunk 1 above)
Comment (tomwilkie) [reply pkg/ingester/index.go]: Actually broke them out into separate functions, as the locking was wrong.

File: pkg/ingester/series_map.go [hunk 2]
```diff
@@ -2,67 +2,77 @@ package ingester
...
+	ms, ok := shard.m[fp]
+	shard.mtx.Unlock()
+	return ms, ok
 }
 
 // put adds a mapping to the seriesMap. It panics if s == nil.
 func (sm *seriesMap) put(fp model.Fingerprint, s *memorySeries) {
-	sm.mtx.Lock()
-	defer sm.mtx.Unlock()
+	shard := &sm.shards[hashFP(fp)%seriesMapShards]
```
Comment (csmarchbanks) [inline pkg/ingester/series_map.go]: This function no longer panics, do we still want it to? Or else update the comment

File: pkg/ingester/series_map.go [hunk 3]
```diff
@@ -2,67 +2,77 @@ package ingester
...
 func (sm *seriesMap) del(fp model.Fingerprint) {
-	sm.mtx.Lock()
-	defer sm.mtx.Unlock()
-
-	delete(sm.m, fp)
+	shard := &sm.shards[hashFP(fp)%seriesMapShards]
+	shard.mtx.Lock()
+	delete(shard.m, fp)
+	shard.mtx.Unlock()
+	atomic.AddInt32(&sm.size, -1)
```
Comment (csmarchbanks) [inline pkg/ingester/series_map.go]: What if you try to delete something that doesn't exist? Do you still want to decrement?

File: pkg/ingester/series_map.go (hunk 2 above)
Comment (tomwilkie) [reply pkg/ingester/series_map.go]: Yeah I don't think we want that; will update comment.

**Comment by tomwilkie** [comment]
Thanks for the feedback @csmarchbanks!

File: pkg/ingester/series_map.go (hunk 3 above)
Comment (tomwilkie) [reply pkg/ingester/series_map.go]: Don't think it'll ever happen, but it doesn't hurt to protect against it.

**Pull request merged**